benchmarks/sparsereg_run.sh

```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+set -xe
+
+device=cpu
+N=100
+benchmark_dir=$( cd $(dirname "$0") ; pwd -P )
+
+for backend in numpyro stan; do
+    for seed in 1 2 3 4 5; do
+        for P in 10 20 30 40 50 60 70 80; do
+            if [[ ${device} = "cpu" ]]; then
+                python ${benchmark_dir}/sparse_regression.py --backend ${backend} --num-data ${N} \
+                    --num-dimensions ${P} --device ${device} --seed ${seed} --x64 #--disable-progbar
```

Comment:
@fehiepsi - I have commented out `disable-progbar` because currently `fori_collect` incurs significant compilation cost and is slower. I'm finding that results are faster and better with progress bar enabled. I wrote a test to check if we get the same results with / without progbar, which seems to pass for a simple example. However for this example (sparse regression), the warmup state and the overall results are different. I'm not sure why that should be the case.

```python
def test_progbar():
    true_mean, true_std = 1., 2.
    num_warmup, num_samples = 10, 80

    def model(data):
        mean = numpyro.param('mean', 0.)
        std = numpyro.param('std', 1., constraint=constraints.positive)
        return numpyro.sample('obs', dist.Normal(mean, std), obs=data)

    data = dist.Normal(true_mean, true_std).sample(random.PRNGKey(1), (2000,))
    kernel = NUTS(model=model)
    # num_samples must be >= 1; otherwise we'll get an error
    mcmc = MCMC(kernel, num_warmup)
    mcmc.run(random.PRNGKey(2), data)
    mcmc.num_samples = num_samples
    mcmc.run(random.PRNGKey(3), data, reuse_warmup_state=True)
    mcmc1 = MCMC(kernel, num_warmup, progress_bar=False)
    t1 = time.time()
    mcmc1.run(random.PRNGKey(2), data)
    t2 = time.time()
    print(t2 - t1)
    mcmc1.num_samples = num_samples
    mcmc1.run(random.PRNGKey(3), data, reuse_warmup_state=True)
    t3 = time.time()
    print(t3 - t2)
    check_close(mcmc1.get_samples(), mcmc.get_samples(), atol=1e-3, rtol=0.05)
    check_close(mcmc1._warmup_state, mcmc._warmup_state, atol=1e-3, rtol=0.05)
```

Reply:
I have committed the test, in case you want to play with it, but I'll remove it eventually.

Reply:
Yup, I expected that the second run will compile everything again. For example,
```python
import time
import jax; jax.config.update("jax_platform_name", "cpu")

def body_fn(i, val):
    key, x = val
    key, subkey = jax.random.split(key)
    return key, x + jax.random.gamma(subkey, 1.)

tic = time.time()
body_fn(0, (jax.random.PRNGKey(0), 1.))[0].copy()
print(time.time() - tic)

tic = time.time()
body_fn(0, (jax.random.PRNGKey(0), 1.))[0].copy()
print(time.time() - tic)

tic = time.time()
jax.lax.fori_loop(0, 10, body_fn, (jax.random.PRNGKey(0), 1.))[0].copy()
print(time.time() - tic)

tic = time.time()
jax.lax.fori_loop(0, 20, body_fn, (jax.random.PRNGKey(0), 1.))[0].copy()
print(time.time() - tic)
```
would print
```
1.6252799034118652
0.0012903213500976562
1.8120296001434326
0.0021545886993408203
```

Re test_progbar: it passes in my system (jax 0.1.53 and jaxlib 0.1.36)...

I'll try to see if we can trigger the cache when progbar=False and `collection` has the same shape.

Reply:
In your benchmark the only thing that I find surprising  is that even if a part (or most) of the `body_fn` is pre-compiled, e.g. 

```python
@jax.jit
def sample_gamma(key, x):
    key, subkey = jax.random.split(key)
    return key, jax.random.gamma(subkey, 1.) + x


def body_fn(i, val):
    key, x = val
    return sample_gamma(key, x)
```

it will still take a good amount of time to compile the `fori_loop` - precompiling significant portion of the `body_fn` doesn't mean that the compilation of the `fori_loop` will be significantly faster. Once we realize that, its probably not surprising what we observe.

> I'll try to see if we can trigger the cache when progbar=False and collection has the same shape.

That shouldn't be hard to do (I can put up something if that seems important), but I'm not sure if it is worth doing that since this will only matter on multiple runs. I am much more concerned about correctness, that is the fact that we get quite different answers on this example seems fishy.

Reply:
I think this has been observed in neutra example too: fori_loop/scan, for loop with jit(svi.update), fast_math enable/disable gives different answers. I guessed the issue happens at `cholesky` at that time but haven't looked back for it...

Reply:
> Once we realize that

I meant so (`jax.random.gamma` is jitted by default). :)

> this will only matter on multiple runs

I think we supported multiple run with different data for progress_bar=True, so it is worth to support for progress_bar=False too. Actually, besides improving the benchmark, I also need this optimization for parallel hmc, where multi-chain is used (current optimization only applies for 1 chain MCMC). I just learned that `pmap` can use the cached
```
import time
import jax
import numpyro; numpyro.set_host_device_count(10)

def sample_gamma(key):
    return jax.random.gamma(key, 1.)

tic = time.time()
jax.pmap(sample_gamma)(jax.random.split(jax.random.PRNGKey(0), 10))
print(time.time() - tic)

tic = time.time()
jax.pmap(sample_gamma)(jax.random.split(jax.random.PRNGKey(1), 10))
print(time.time() - tic)
```
I have been trying to store `_body_fn` in `fori_collect`, WDYT about doing so? If you have a different idea, please let me know.

> I can put up something if that seems important

This would be great! :D

Reply:
To support this, I think we need to make a couple of changes:
- Not use fori_collect in MCMC, rewrite it such that _body_fn is cached inside MCMC.
- Drop the pattern: num_samples=0... num_warmup and num_samples are fixed values.
- Create a (hidden) method: run_warmup to only run the warmup phase, store the last value in a hidden warmup_state.

WDYT?

Reply:
That sounds reasonable, but I think we can support the full pattern with some small changes. I do agree that we shouldn't advertise this pattern and maybe introduce a `_` before these parameters, so as to not encourage using it, but it will be nice to be able to support it.

Reply:
Yeah, let me try to make hidden methods `_compile` (to separate compiling time), `_warmup` (to separate warmup time),... to test out if caching works (this is not possible without your recent PRs). But feel free to make your changes, so we can have different versions to discuss/choose from.

Reply:
Could you take a look at https://github.com/pyro-ppl/numpyro/pull/479? I think that's simpler than other alternatives? Feel free to push to it directly, I'll take a look at it again in the afternoon.